lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -265,6 +266,64 @@ VariableOp::handlePromotionComplete(const MemorySlot &slot, Value defaultValue,
   return std::nullopt;
 }
 
+SmallVector<DestructurableMemorySlot> VariableOp::getDestructurableSlots() {
+  auto refType = getType();
+  auto destructurable = llvm::dyn_cast<DestructurableTypeInterface>(refType);
+  if (!destructurable)
+    return {};
+
+  auto destructuredType = destructurable.getSubelementIndexMap();
+  if (!destructuredType)
+    return {};
+
+  return {DestructurableMemorySlot{{getResult(), refType}, *destructuredType}};
+}
+
+DenseMap<Attribute, MemorySlot> VariableOp::destructure(
+    const DestructurableMemorySlot &slot,
+    const SmallPtrSetImpl<Attribute> &usedIndices, OpBuilder &builder,
+    SmallVectorImpl<DestructurableAllocationOpInterface> &newAllocators) {
+  builder.setInsertionPointAfter(*this);
+
+  DenseMap<Attribute, MemorySlot> slotMap;
+  auto destructurable = llvm::cast<DestructurableTypeInterface>(getType());
+  llvm::ArrayRef<StructLikeMember> members;
+  TypeSwitch<Type>(getType().getNestedType())
+      .Case<StructType, UnpackedStructType>(
+          [&members](auto &type) { members = type.getMembers(); })
+      .Default([this](auto &) {
+        emitOpError("Result type must be StructType or UnpackedStructType");
+      });
+  SmallVector<Value> inputs;
+  for (Attribute usedIndex : usedIndices) {
+    auto elemType =
+        cast<UnpackedType>(destructurable.getTypeAtIndex(usedIndex));
+    auto elemRefType = RefType::get(elemType);
+    StringAttr name = {};
+    for (const auto &member : members) {
+      if (member.type == elemType)
+        name = member.name;
+    }
+    auto varOp =
+        builder.create<VariableOp>(getLoc(), elemRefType, name, Value());
+    // context.valueSymbols.insertIntoScope(context.valueSymbols.getCurScope(),&var,
+    // varOp);
+    inputs.push_back(varOp);
+    slotMap.try_emplace<MemorySlot>(usedIndex, {varOp, elemType});
+  }
+  builder.create<StructCreateOp>(getLoc(), getType(), inputs);
+
+  return slotMap;
+}
+
```

Comment:
@Moxinilian I just support DestructurableAllocationOpInterface to variableOp, but it seems no working for struct. Could you have a look, thanks a lot.